### PR TITLE
Calculate `base_score` based on input labels.

### PR DESCRIPTION
**Pull request by trivialfis**
https://github.com/dmlc/xgboost/issues/4321 .

This PR calculates the `base_score` from labels for l1 regression and saves it to the output model. Will follow up on other objectives as well.
- Configure model parameter for base_score.
- Add estimation function in objective.
- Change base_score to an array. At the moment, we use only 1 element, the change is to prepare for multi-class and multi-output once we can remove some legacy in the binary model.

Multi-target and multi-class are not yet supported due to the binary model parameter.

### CODE REVIEW

**Comment by trivialfis** [comment]
The old binary format and the `gpu_id` configuration are probably too difficult to workaround for this PR.

File: src/objective/objective.cc [hunk 1]
```diff
@@ -31,6 +31,11 @@ ObjFunction* ObjFunction::Create(const std::string& name, GenericParameter const
   return pobj;
 }
 
+void ObjFunction::InitEstimation(MetaInfo const&, linalg::Tensor<float, 1>* base_score) const {
```
Comment (RAMitchell) [inline src/objective/objective.cc:34]: So, at this time, default behaviour will be to use the initialisation for absolute error but not other loss functions? Should we add this for all loss functions? I think it makes sense, but we might need to manage user expectations when their model accuracy changes substantially between versions.

File: src/learner.cc [hunk 2]
```diff
@@ -308,8 +353,58 @@ class LearnerConfiguration : public Learner {
...
+    } else {
+      // lastly, if data is not available (prediction for custom objective), use default.
+      base_score.Reshape(1);
+      base_score(0) = ObjFunction::DefaultBaseScore();
+    }
+
+    auto task = obj_->Task();
+    auto in = base_score.HostView();
+    rabit::Allreduce<rabit::op::Sum>(in.Values().data(), in.Values().size());
+    std::transform(linalg::cbegin(in), linalg::cend(in), linalg::begin(in),
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/learner.cc:394]: Maybe the averaging of the base score needs to be handled in InitEstimation, on a case by case basis for each objective.

File: src/learner.cc [hunk 3]
```diff
@@ -308,8 +353,58 @@ class LearnerConfiguration : public Learner {
...
+    // Under what condition should we omit the transformation:
+    //
+    // - base_score is loaded from old binary model.
+    //
+    // What are the other possible conditions:
+    //
+    // - model loaded from new binary or JSON.
+    // - model is created from scratch.
+    // - model is configured second time due to change of parameter
+    CHECK(obj_);
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/learner.cc:376]: I agree.  That's why I really want to remove the old model format.

File: src/learner.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/learner.cc:394]: Will change the average to be more robust.

File: src/objective/objective.cc (hunk 1 above)
Comment (trivialfis) [reply src/objective/objective.cc:34]: I plan to work on other functions as well.  This PR is just to get started with basic stuff like model serialization.  The biggest issue for now is multi class model where we need a vector of base scores.

**Comment by trivialfis** [comment]
I removed the use of `nan` as the base score flag to avoid breaking changes in downstream libraries (like treelite). Instead, a new `base_score_estimated` variable is introduced, but the variable is not read during model load so we don't need to keep it stable.

File: src/objective/objective.cc (hunk 1 above)
Comment (trivialfis) [reply src/objective/objective.cc:34]: I have a prototype for least square. Will follow up once this PR is sorted out.

File: src/objective/regression_obj.cu [hunk 4]
```diff
@@ -698,6 +703,28 @@ class MeanAbsoluteError : public ObjFunction {
...
+    }
+
+    auto world = static_cast<float>(rabit::GetWorldSize());
+    rabit::Allreduce<rabit::op::Sum>(&invalid, 1);  // number of empty workers
+    world -= static_cast<float>(invalid);           // number of non-empty workers
+
+    // average base score across all valid workers
+    rabit::Allreduce<rabit::op::Sum>(out.Values().data(), out.Values().size());
+    std::transform(linalg::cbegin(out), linalg::cend(out), linalg::begin(out),
+                   [world](float v) { return v / world; });
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/objective/regression_obj.cu]: Thank you for the suggestion, changed it to the weighted average. I have adapted your example into a Python test.

**Review by RAMitchell — APPROVED** [review]
